File: Bluetooth/c/test-bth1208LS.c
/*
 *  Copyright (c) 2016 Warren J. Jasper <wjasper@ncsu.edu>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <stdbool.h>
#include <fcntl.h>
#include <math.h>
#include <ctype.h>
#include <time.h>
#include <sys/socket.h>
#include <sys/select.h>
#include <sys/time.h>
#include <sys/types.h>
#include <bluetooth/bluetooth.h>
#include <bluetooth/rfcomm.h>
#include "bth-1208LS.h"

/* Test Program */
int toContinue()
{
  int answer;
  answer = 0; //answer = getchar();
  printf("Continue [yY]? ");
  while((answer = getchar()) == '\0' ||
    answer == '\n');
  return ( answer == 'y' || answer == 'Y');
}

int main(int argc, char**argv)
{
  struct tm calDate;
  DeviceInfo_BTH1208LS device_info;
  uint8_t options;
  uint8_t channel, channels;
  int flag;
  int ch;
  int i, j, k, m;
  char serial[9];  // serial number
  uint16_t status, version, radioVersion;
  uint8_t range;
  uint8_t ranges[4] = {0, 0, 0, 0};
  uint16_t value;
  uint16_t values[2];
  uint32_t count;
  double frequency, voltage;
  uint16_t dataAIn[4*12192];  // holds 16 bit unsigned analog input data
  uint16_t data;
  int nChan, repeats, nScan;

  memset(&device_info, 0x0, sizeof(device_info));
  device_info.device.frameID = 0;         // zero out the frameID

  if (argc == 2) {
    strncpy(device_info.device.baddr, argv[1], 18);
    printf("BTH-1208LS BADDR address = %s\n", argv[1]);
  } else if (discoverDevice(&device_info.device, "BTH-1208LS-6833") < 0) {
    printf("No device found.\n");
    return -1;
  }

  if (openDevice(&device_info.device) < 0) {
    printf("Error openeing device\n");
    exit(-1);
  } else {
    printf("Found a Bluetooth BTH-1208LS.\n");
  }

  // some initialization
  BuildGainTable_DE_BTH1208LS(&device_info);
  BuildGainTable_SE_BTH1208LS(&device_info);
  for (i = 0; i < NCHAN_DE; i++) {   // channel
    for (j = 0; j < NGAINS; j++ ) {      // range 
      printf("Calibration Table (Differential): Channel = %d  Gain = %d  Slope = %f   Offset = %f\n", 
	     i, j, device_info.table_AInDE[i][j].slope, device_info.table_AInDE[i][j].intercept);
    }
  }
  printf("\n");
  for (i = 0; i < NCHAN_SE; i++ ) {  // channel  (SE only has 1 gain range)
    printf("Calibration Table (Single Ended): Channel = %d  Slope = %f   Offset = %f\n", 
	   i, device_info.table_AInSE[i].slope, device_info.table_AInSE[i].intercept);
  }

  CalDate_BTH1208LS(&device_info, &calDate);
  printf("\n");
  printf("MFG Calibration date = %s\n", asctime(&calDate));

  while(1) {
    printf("\n\nBTH-1208LS Testing\n");
    printf("----------------\n");
    printf("Hit 'b' to blink\n");
    printf("Hit 'c' to test counter\n");
    printf("Hit 'C' to for continuous sampling\n");
    printf("Hit 'd' to test digitial IO\n");
    printf("Hit 'i' to test Analog Input\n");
    printf("Hit 'I' to test Analog Input Scan\n");
    printf("Hit 'o' to test Analog Output\n");
    printf("Hit 'x' to test Analog Input Scan (Multi-channel)\n");
    printf("Hit 'r' to reset the device\n");
    printf("Hit 's' to get serial number\n");
    printf("Hit 'S' to get Status\n");
    printf("Hit 'v' to get the battery voltage in mV\n");
    printf("Hit 'e' to exit\n");

    while((ch = getchar()) == '\0' || ch == '\n');
    switch(ch) {
      case 'b': /* test to see if LED blinks */
        printf("Enter number or times to blink: ");
        scanf("%hhd", &options);
	BlinkLED_BTH1208LS(&device_info, options);
        break;
      case 'c':
        ResetCounter_BTH1208LS(&device_info);
        printf("Connect AO 0 to CTR.\n");
        toContinue();
        for (i = 0; i < 100; i++) {                   // toggle
    	  AOut_BTH1208LS(&device_info, 0, 4095);
	  AOut_BTH1208LS(&device_info, 0, 0);
        }
	Counter_BTH1208LS(&device_info, &count);
        printf("Count = %d.  Should read 100.\n", count);
        break;
      case 'd':
	printf("Test of DIO Out\n");
	printf("Input value [0-ff]: ");
	scanf("%hhx", &options);
	DOut_BTH1208LS(&device_info, options);
	DOutR_BTH1208LS(&device_info, &options);
	printf("The value you wrote = %#x", options);
	break;
      case 'e':
        close(device_info.device.sock);
	printf("Success!\n");
	return 0;
	break;
      case 'i':
	printf("Input channel DE [0-3]: ");
	scanf("%hhd", &channel);
        printf("Input range [0-7]: ");
	scanf("%hhd", &range);
	for (i = 0; i < 20; i++) {
	  AIn_BTH1208LS(&device_info, channel, DIFFERENTIAL, range, &value);
	  value = rint(value*device_info.table_AInDE[channel][range].slope + device_info.table_AInDE[channel][range].intercept);
  	  printf("Range %d  Channel %d   Sample[%d] = %#x Volts = %lf\n",
		   range, channel,  i, value, volts_BTH1208LS(value, range));
	  usleep(50000);	  
	}
        break;
      case 'I':
	printf("Testing BTH-1208lS Analog Input Scan.\n");
        printf("Enter number of scans: ");
        scanf("%d", &count);
	printf("Input channel 0-3: ");
        scanf("%hhd", &channel);
        printf("Enter sampling frequency [Hz]: ");
	scanf("%lf", &frequency);
        printf("Enter Range [0-7]: ");
        scanf("%hhd", &range);
        ranges[channel] = range;
        options = DIFFERENTIAL_MODE;
	AInScanStop_BTH1208LS(&device_info);
	AInScanClearFIFO_BTH1208LS(&device_info);
        AInConfigW_BTH1208LS(&device_info, ranges);
	memset(dataAIn, 0x0, sizeof(dataAIn));
	AInConfigR_BTH1208LS(&device_info, ranges);
	AInScanStart_BTH1208LS(&device_info, count, 0x0, frequency, (0x1<<channel), options);
	AInScanRead_BTH1208LS(&device_info, count, dataAIn);
	for (i = 0; i < count; i++) {
	  dataAIn[i] = rint(dataAIn[i]*device_info.table_AInDE[channel][range].slope + device_info.table_AInDE[channel][range].intercept);
          printf("Range %d Channel %d  Sample[%d] = %#x Volts = %lf\n", range, channel,
		 i, dataAIn[i], volts_BTH1208LS(dataAIn[i], range));
	}
        break;
      case 'x':
        printf("Testing BTH-1208LS Multi-Channel Analog Input Scan.\n");
        printf("Enter number of channels (1-4): ");
        scanf("%d", &nChan);
        printf("Enter number of scans: ");
        scanf("%d", &nScan);
        printf("Enter number of repeats: ");
        scanf("%d", &repeats);
	printf("Enter sampling frequency: ");
        scanf("%lf", &frequency);
        // Build bitmap for the first nChan in channels.
        channels = 0;
        for (i = 0; i < nChan; i++) {
	  channels |= (1 << i);
	}
        // Always use BP_20V to make it easy (BP_20V is 0...)
        range = BP_20V;
        memset(ranges, BP_20V, sizeof(ranges));
	AInScanStop_BTH1208LS(&device_info);
	AInScanClearFIFO_BTH1208LS(&device_info);
        AInConfigW_BTH1208LS(&device_info, ranges);
        options = DIFFERENTIAL_MODE;
        // Run a loop for the specified number of repeats and
        // show the results...
        for (m = 0; m < repeats; m++) {
	  printf("\n\n---------------------------------------");
	  printf("\nrepeat: %d\n", m);
	  AInScanStop_BTH1208LS(&device_info);
	  AInScanClearFIFO_BTH1208LS(&device_info);
	  AInScanStart_BTH1208LS(&device_info, nScan*nChan, 0x0, frequency, channels, options);
	  AInScanRead_BTH1208LS(&device_info, nScan, dataAIn);
	  for (i = 0; i < nScan; i++) {
	    printf("%6d", i);
	    for (j = 0; j < nChan; j++)	{
	      k = i*nChan + j;
	      data = rint(dataAIn[k]*device_info.table_AInDE[j][range].slope + device_info.table_AInDE[j][range].intercept);
	      printf(", %8.4f", volts_BTH1208LS(data, range));
	    } /* for (j - 0; j < 8, j++) */
	    printf("\n");
	  } /* for (i = 0; i < nScan; i++) */
	} /* for (m = 0; m < repeats; m++) */
	printf("\n\n---------------------------------------");
	break;
      case 'C':
        printf("Testing BTH-1208LS Continuous Analog Input Scan.\n");
	printf("Hit any key to exit\n");
	printf("Enter sampling frequency: ");
        scanf("%lf", &frequency);
	channels = 0x2;
	range = BP_20V;
	memset(ranges, BP_20V, sizeof(ranges));
	AInScanStop_BTH1208LS(&device_info);
	AInScanClearFIFO_BTH1208LS(&device_info);
        AInConfigW_BTH1208LS(&device_info, ranges);
        options = DIFFERENTIAL_MODE;
	nScan = 0;    // Continuous scan
	flag = fcntl(fileno(stdin), F_GETFL);
	fcntl(0, F_SETFL, flag | O_NONBLOCK);
	AInScanStart_BTH1208LS(&device_info, nScan, 0x0, frequency, channels, options);
	i = 0;
	do {
	  device_info.nDelay = (127.*1000.)/frequency;     // delay in ms
	  usleep(device_info.nDelay*900);                  // sleep in us
	  j = AInScanSendData_BTH1208LS(&device_info, 127, dataAIn, device_info.nDelay);
	  printf("Scan = %d, samples returned = %d\n", i, j);
	  i++;
	} while (!isalpha(getchar()));
	fcntl(fileno(stdin), F_SETFL, flag);
	AInScanStop_BTH1208LS(&device_info);
	break;
      case 'o':
	printf("Test Analog Output\n");
        printf("Enter Channel [0-1] ");
        scanf("%hhd", &channel);
        printf("Enter voltage [0-2.5V]: ");
	scanf("%lf", &voltage);
        value = voltage * 4095 / 2.5;
	AOut_BTH1208LS(&device_info, channel, value);
	AOutR_BTH1208LS(&device_info, values);
	printf("Analog Output Voltage = %f V\n", volts_BTH1208LS(values[channel], UP_2_5V));
        break;
      case 'S':
	Status_BTH1208LS(&device_info, &status);
	FirmwareVersion_BTH1208LS(&device_info, &version);
	RadioFirmwareVersion_BTH1208LS(&device_info, &radioVersion);
	printf("Status = %#x   Firmware Version = %x.%x   Radio Firmware Version = %x.%x\n",
	       status, (version >> 8) & 0xff, version & 0xff, (radioVersion >> 8) & 0xff, radioVersion & 0xff);
	break;
      case 's':
        GetSerialNumber_BTH1208LS(&device_info, serial);
        printf("Serial number = %s\n", serial);
        break;
      case 'r':
        Reset_BTH1208LS(&device_info);
        break;
      case 'v':
        BatteryVoltage_BTH1208LS(&device_info);
        printf("Battery Voltage = %d mV\n", device_info.voltage);
	break;
    }
  }
}

File: Bluetooth/python/test-bth1208LS.py
#! /usr/bin/python3
#
# Copyright (c) 2019 Warren J. Jasper <wjasper@ncsu.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

from bth_1208LS import *
import time
import sys
import fcntl
import os

def toContinue():
  answer = input('Continue [yY]? ')
  if (answer == 'y' or answer == 'Y'):
    return True
  else:
    return False

def main():
  target_name = "BTH-1208LS-6833"
  device = []

  if len(sys.argv) == 2:
    device.append(mccBluetoothDevice(sys.argv[1]))
  else:
    # Discover a BTH-1208LS device
    device.append(mccBluetoothDevice(discoverDevice(target_name)))

  if (len(device) > 0):
    print('Number of devices found = ', len(device))
  else:
    print('No device', target_name, 'found')
    exit(0)

  # Open the device
  try:
    device[0].openDevice()
  except:
    print("Can not open device. Could be in charging mode.")
    exit(0)

  # initalize the class
  bth1208LS = BTH_1208LS(device[0])

  # allow charging with Bluetooth connected.
  data = [1]
  bth1208LS.SettingsMemoryW(0xe, 1, data)

  # print out the settings memory
  address = 0x0
  print("\nDumping Settings Memory: ")
  print(bth1208LS.SettingsMemoryR(address, 15))
  print("")

  # print out the calibration tables:
  for chan in range(bth1208LS.NCHAN_DE):
    for gain in range(bth1208LS.NGAINS):
      print('Calibration Table (Differential): Chan =',chan,' Range = ',gain, \
            'Slope = ',format(bth1208LS.table_AInDE[chan][gain].slope,'.5f'),\
            'Intercept = ',format(bth1208LS.table_AInDE[chan][gain].intercept,'5f'))

  print("")
  for chan in range(bth1208LS.NCHAN_SE):
    print('Calibration Table (Single Ended): Chan = ', chan, 
          'Slope = ',format(bth1208LS.table_AInSE[chan].slope,'.5f'),\
          'Intercept = ',format(bth1208LS.table_AInSE[chan].intercept,'5f'))

  # print last known calibration date:
  mdate = bth1208LS.CalDate()
  print('\nMFG Calibration date: ', mdate)

  while True:
    print("\nBTH-1208LS Testing")
    print("----------------")
    print("Hit 'b' to blink.")
    print("Hit 'c' to test counter.")
    print("Hit 'C' for continuous sampling")
    print("Hit 'd' to read digital IO.")
    print("Hit 'D' to write digital IO.")
    print("Hit 'i' to test Analog Input")
    print("Hit 'I' to test Analog Input Scan")
    print("Hit 'o' to test Analog Output")
    print("Hit 'x' to test Analog Input Scan (Multi-channel)")
    print("Hit 'r' to reset the device.")
    print("Hit 'e' to exit.")
    print("Hit 's' to get serial number.")
    print("Hit 'S' to get Status.")
    print("Hit 'v' to get the battery voltage in mV.")

    ch = input('\n')
    if ch == 'b':
      count = int(input('Enter number of times to blink: '))
      bth1208LS.BlinkLED(count)
    elif ch == 'c':
      bth1208LS.ResetCounter()
      print("Connect AO 0 to CTR.")
      toContinue()
      for i in range(100):
        bth1208LS.AOut(0, 4095)
        bth1208LS.AOut(0, 0)
      count = bth1208LS.Counter()
      print("Count = ", count, "    Should read 100.")
    elif ch == 'd':
      value = bth1208LS.DIn()
      print("Digital IO pins: ", hex(value))
    elif ch == 'D':
      print("Test of DIO Out")
      value = int(input("Input value [0-ff]: "),base=16) & 0xff
      bth1208LS.DOut(value)
      value = bth1208LS.DOutR()
      print("The value you wrote = ", hex(value))
    elif ch == 'i':
      channel = int(input("Input channel DE [0-3]: "))
      gain = int(input("Input range [0-7]: "))
      mode = bth1208LS.DIFFERENTIAL
      for i in range(20):
        value = bth1208LS.AIn(channel, mode, gain)
        print("Range = {0:d},  Channel = {1:d},  Sample[{2:d}] = {3:x}    Volts = {4:f}"\
              .format(gain, channel, i, value, bth1208LS.volts(value, gain)))
    elif ch == 'I':
      print("Testing BTH-1208LS Analog Input Scan.")
      count = int(input("Enter number of scans: "))
      channel = int(input("Input channel DE [0-3]: "))
      gain = int(input("Input range [0-7]: "))
      frequency = float(input("Enter sampling frequency [Hz]: "))
      ranges = [0x0, 0x0, 0x0, 0x0]
      ranges[channel] = gain
      options = bth1208LS.DIFFERENTIAL_MODE
      bth1208LS.AInScanStop()
      bth1208LS.AInScanClearFIFO()
      bth1208LS.AInConfigW(ranges)
      ranges = bth1208LS.AInConfigR()
      bth1208LS.AInScanStart(count, 0x0, frequency, (0x1 << channel), options)
      dataAIn = bth1208LS.AInScanRead(count)
      for i in range(count):
        dataAIn[i] = round(dataAIn[i]*bth1208LS.table_AInDE[channel][gain].slope + bth1208LS.table_AInDE[channel][gain].intercept)
        print("Range {0:d}  Channel {1:d}  Sample[{2:d}] = ".format(gain, channel, i), hex(dataAIn[i])," Volts = {0:7.4f}".format(bth1208LS.volts(dataAIn[i], gain)))
    elif ch == 'x':
      print("Testing BTH-1208LS Multi-Channel Analog Input Scan.")
      nChan = int(input("Enter number of channels (1-4): "))
      nScan = int(input("Enter number of scans: "))
      repeats = int(input("Enter number of repeats: "))
      frequency = float(input("Enter sampling frequency [Hz]: "))
      # build bitmap for the first nchan in channels
      channels = 0
      for i in range(nChan):
        channels |= (0x1 << i)
      # Always use BP_20V to make it easy (BP_20V is 0...)
      gain = bth1208LS.BP_20V
      ranges = [0x0, 0x0, 0x0, 0x0]
      for i in range(4):
        ranges[i] = gain
      options = bth1208LS.DIFFERENTIAL_MODE
      bth1208LS.AInScanStop()
      bth1208LS.AInScanClearFIFO()
      bth1208LS.AInConfigW(ranges)
      time.sleep(1)
      for m in range(repeats):
        print("\n\n---------------------------------------");
        print("\nrepeat: ", m);
        bth1208LS.AInScanStop()
        bth1208LS.AInScanClearFIFO()
        bth1208LS.AInScanStart(nScan*nChan, 0x0, frequency, channels, options)
        data = bth1208LS.AInScanRead(nScan)
        for i in range(int(nScan)):
          print("{0:6d}".format(i), end="")
          for j in range(nChan):
            k = i*nChan + j
            value = round(data[k]*bth1208LS.table_AInDE[j][gain].slope + bth1208LS.table_AInDE[j][gain].intercept)
            print(" ,{0:8.4f}".format(bth1208LS.volts(value, ranges[0])), end="")
          print("")
        bth1208LS.AInScanStop()
        bth1208LS.AInScanClearFIFO()
    elif ch == 'C':
      print('BTH-1208LS Continuous Samping')
      print('Hit space <CR> to exit.')
      frequency = float(input('Enter desired frequency [Hz]: '))
      options = bth1208LS.DIFFERENTIAL_MODE  # Single Ended mode
      count = 0                              # continuous execution
      channels = 0x01                        # read all 8 channels
      bth1208LS.AInScanStop()
      bth1208LS.AInScanClearFIFO()
      bth1208LS.AInScanStart(count, 0x0, frequency, channels, options)
      flag = fcntl.fcntl(sys.stdin, fcntl.F_GETFL)
      fcntl.fcntl(sys.stdin, fcntl.F_SETFL, flag|os.O_NONBLOCK)
      j = 0
      while True:
        raw_data = bth1208LS.AInScanRead(127)
        print('Scan =', j, 'samples returned =', len(raw_data))
        j += 1
        c = sys.stdin.readlines()
        if (len(c) != 0):
          break
      fcntl.fcntl(sys.stdin, fcntl.F_SETFL, flag)
      bth1208LS.AInScanStop()
    elif ch == 'e':
      bth1208LS.device.sock.close()
      exit(0)
    elif ch == 's':
      print('Serial Number:',bth1208LS.GetSerialNumber())
    elif ch == 'S':
      status = bth1208LS.Status()
      version = bth1208LS.FirmwareVersion()
      radioVersion = bth1208LS.RadioFirmwareVersion()
      print("Status:", hex(status), "    Firmware Version: {0:1x}.{1:2x}    Radio FirmwareVersion: {2:1x}.{3:2x}"\
            .format ((version >> 8) & 0xff, (version & 0xff), (radioVersion >> 8) & 0xff, (radioVersion & 0xff)))
    elif ch == 'r':
      bth1208LS.Reset()
    elif ch == 'v':
      voltage = bth1208LS.BatteryVoltage()
      print("Battery Voltage {0:d} mV".format(voltage))
    elif ch == 'o':
      print("Test Analog Output")
      channel = int(input("Enter Channel [0-1]: "))
      voltage = float(input("Enter voltage [0-2.5V]: "))
      value = int(voltage * 4095 / 2.5)
      bth1208LS.AOut(channel, value)
      print("Analog Output Voltage = ", bth1208LS.AOutR()[channel]*2.5/4095)

if __name__ == "__main__":
  main()
